cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i

```diff
@@ -458,74 +726,211 @@ void OrderedRespHelper<algorithmFPType, cpu>::computeHistWithoutWeights(algorith
+    auto nFeatIdx    = this->_idxFeatureBuf.get(); //number of indexed feature values, array
 
     intermSummFPType bestImpDecreasePart =
         split.impurityDecrease < 0 ? -1 : (split.impurityDecrease + curImpurity.mean * curImpurity.mean) * totalWeights;
     size_t nLeft                = 0;
     algorithmFPType leftWeights = 0.;
     intermSummFPType sumLeft    = 0;
     int idxFeatureBestSplit     = -1; //index of best feature value in the array of sorted feature values
-    for (size_t i = 0; i < nDiffFeatMax; ++i)
+
+    size_t minidx = 0;
+    size_t maxidx = nDiffFeatMax - 1;
+
+    while ((minidx < maxidx) && isZero<IndexType, cpu>(nFeatIdx[minidx])) minidx++;
+
+    while ((minidx < maxidx) && isZero<IndexType, cpu>(nFeatIdx[maxidx])) maxidx--;
+
+    DAAL_ASSERT(minidx < maxidx); //if the if statement after minidx search doesn't activate, we have an issue.
+    if (minidx == maxidx)
     {
-        if (!nFeatIdx[i]) continue;
+        return idxFeatureBestSplit;
+    }
 
-        algorithmFPType thisFeatWeights = noWeights ? nFeatIdx[i] : featWeights[i];
+    //randomly select a histogram split index
+    size_t idx;
+    RNGs<size_t, cpu> rng;
+    rng.uniform(1, &idx, this->engineImpl->getState(), minidx, maxidx); //find random index between minidx and maxidx
 
-        nLeft       = (featureUnordered ? nFeatIdx[i] : nLeft + nFeatIdx[i]);
-        leftWeights = (featureUnordered ? thisFeatWeights : leftWeights + thisFeatWeights);
-        if ((nLeft == n) //last split
-            || ((n - nLeft) < nMinSplitPart) || ((totalWeights - leftWeights) < minWeightLeaf))
-            break;
-        sumLeft = (featureUnordered ? buf[i] : sumLeft + buf[i]);
-        if ((nLeft < nMinSplitPart) || (leftWeights < minWeightLeaf)) continue;
+    while (isZero<IndexType, cpu>(nFeatIdx[idx]))
+    {
+        idx--;
+    }
+
+    if (noWeights)
+    {
+        if (featureUnordered)
+        {
+            nLeft       = nFeatIdx[idx];
+            sumLeft     = buf[idx];
+            leftWeights = nFeatIdx[idx];
+        }
+        else
+        {
+            PRAGMA_IVDEP
+            PRAGMA_VECTOR_ALWAYS
+            for (size_t i = minidx; i <= idx; ++i)
+            {
+                nLeft += nFeatIdx[i];
+                sumLeft += buf[i];
+            }
+            leftWeights = nLeft;
+        }
+    }
+    else
+    {
+        if (featureUnordered)
+        {
+            nLeft       = nFeatIdx[idx];
+            sumLeft     = buf[idx];
+            leftWeights = featWeights[idx];
+        }
+        else
+        {
+            PRAGMA_IVDEP
+            PRAGMA_VECTOR_ALWAYS
+            for (size_t i = minidx; i <= idx; ++i)
+            {
+                nLeft += nFeatIdx[i];
+                sumLeft += buf[i];
+                leftWeights += featWeights[i];
+            }
+        }
+    }
+
+    if (!(((n - nLeft) < nMinSplitPart) || ((totalWeights - leftWeights) < minWeightLeaf) || (nLeft < nMinSplitPart)
```

Comment:
Problem with original code, its either original or a copy over for similarity purposes. I would favor a refactoring of the code, rather than small changes...